pkg/pool-weighted/contracts/smart/IndexPool.sol

```diff
@@ -97,20 +113,140 @@ contract IndexPool is BaseWeightedPool {
+            return 0;
+        }
+
+        // No need for SafeMath as it was checked right above: endTime >= currentTime >= startTime
+        uint256 totalSeconds = endTime - startTime;
+        uint256 secondsElapsed = currentTime - startTime;
+
+        // In the degenerate case of a zero duration change, consider it completed (and avoid division by zero)
+        return totalSeconds == 0 ? FixedPoint.ONE : secondsElapsed.divDown(totalSeconds);
+    }
+
+    /**
+    * @dev When calling updateWeightsGradually again during an update, reset the start weights to the current weights,
+    * if necessary.
+    */
+    function _startGradualWeightChange(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory startWeights,
+        uint256[] memory endWeights
+    ) internal virtual {
+        bytes32 newPoolState = _poolState;
+
+        uint256 normalizedSum = 0;
+        for (uint256 i = 0; i < endWeights.length; i++) {
+            uint256 endWeight = endWeights[i];
+            _require(endWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            newPoolState = newPoolState
+            .insertUint31(startWeights[i].compress31(), _START_WEIGHT_OFFSET + i * 31)
+            .insertUint16(endWeight.compress16(), _END_WEIGHT_OFFSET + i * 16);
+
+            normalizedSum = normalizedSum.add(endWeight);
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        _poolState = newPoolState.insertUint32(startTime, _START_TIME_OFFSET).insertUint32(endTime, _END_TIME_OFFSET);
+
+        //        emit GradualWeightUpdateScheduled(startTime, endTime, startWeights, endWeights);
+    }
+
+
+    /**
+     * @dev Schedule a gradual weight change, from the current weights to the given endWeights,
+     * over startTime to endTime
+     */
+//    TODO: has to be authenticate
+    function updateWeightsGradually(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory endWeights
+    ) internal whenNotPaused nonReentrant {
+        InputHelpers.ensureInputLengthMatch(_getTotalTokens(), endWeights.length);
+
+        // If the start time is in the past, "fast forward" to start now
+        // This avoids discontinuities in the weight curve. Otherwise, if you set the start/end times with
+        // only 10% of the period in the future, the weights would immediately jump 90%
+        uint256 currentTime = block.timestamp;
+        startTime = Math.max(currentTime, startTime);
+
+        _require(startTime <= endTime, Errors.GRADUAL_UPDATE_TIME_TRAVEL);
+
+        _startGradualWeightChange(startTime, endTime, _getNormalizedWeights(), endWeights);
+    }
+
+
+    function reweighTokens(address[] calldata tokens, uint256[] calldata desiredWeights) public{
+        uint256 endTime = _poolState.decodeUint32(_END_TIME_OFFSET);
+        require(block.timestamp >= endTime ,"Weight change is already in process");
```

Comment:
Right now it is impossible to start time, + weight change will require a storage somewhere, and it will be complicated to store few weight changes in the byteArray

Reply:
Definitely possible, but it's an overkill for this PR